rviz_rendering/src/rviz_rendering/objects/screw_visual.cpp

```diff
@@ -0,0 +1,164 @@
+ScrewVisual::ScrewVisual(Ogre::SceneManager * scene_manager, Ogre::SceneNode * parent_node)
+{
+  scene_manager_ = scene_manager;
+
+  // Ogre::SceneNode s form a tree, with each node storing the transform (position and orientation)
+  // of itself relative to its parent. Ogre does the math of combining those transforms
+  // for rendering. Here we create a node to store the pose of the screw's header
+  // frame relative to the RViz fixed frame.
+  frame_node_ = parent_node->createChildSceneNode();
+  linear_node_ = frame_node_->createChildSceneNode();
+  angular_node_ = frame_node_->createChildSceneNode();
+  hide_small_values_ = true;
+
+  // We create the arrow object within the frame node so that we can
+  // set its position and direction relative to its header frame.
+  arrow_linear_ = new rviz_rendering::Arrow(scene_manager_, linear_node_);
+  arrow_angular_ = new rviz_rendering::Arrow(scene_manager_, angular_node_);
+  circle_angular_ = new rviz_rendering::BillboardLine(scene_manager_, angular_node_);
+  circle_arrow_angular_ = new rviz_rendering::Arrow(scene_manager_, angular_node_);
+}
+
+ScrewVisual::~ScrewVisual()
+{
+  // Delete the arrow to make it disappear.
+  delete arrow_linear_;
+  delete arrow_angular_;
+  delete circle_angular_;
+  delete circle_arrow_angular_;
+
+  // Destroy the frame node since we don't need it anymore.
+  scene_manager_->destroySceneNode(frame_node_);
+}
+
+void ScrewVisual::setScrew(const Ogre::Vector3 & linear, const Ogre::Vector3 & angular)
+{
+  double linear_length = linear.length() * linear_scale_;
+  double angular_length = angular.length() * angular_scale_;
+  // hide markers if they get too short and hide_small_values_ is activated
+  // "too short" is defined as "linear_length > width_"
+  bool show_linear = (linear_length > width_) || !hide_small_values_;
+  bool show_angular = (angular_length > width_) || !hide_small_values_;
+
+  if (show_linear) {
+    arrow_linear_->setScale(Ogre::Vector3(linear_length, width_, width_));
+    arrow_linear_->setDirection(linear);
+  }
+  linear_node_->setVisible(show_linear);
+
+  if (show_angular) {
+    arrow_angular_->setScale(Ogre::Vector3(angular_length, width_, width_));
+    arrow_angular_->setDirection(angular);
+    Ogre::Vector3 axis_z(0, 0, 1);
+    Ogre::Quaternion orientation = axis_z.getRotationTo(angular);
+    if (std::isnan(orientation.x) || std::isnan(orientation.y) || std::isnan(orientation.z)) {
+      orientation = Ogre::Quaternion::IDENTITY;
+    }
+    // circle_arrow_angular_->setScale(Ogre::Vector3(width_, width_, 0.05));
+    circle_arrow_angular_->set(0, width_ * 0.1, width_ * 0.1 * 1.0, width_ * 0.1 * 2.0);
+    circle_arrow_angular_->setDirection(orientation * Ogre::Vector3(0, 1, 0));
+    circle_arrow_angular_->setPosition(
+      orientation *
+      Ogre::Vector3(angular_length / 4, 0, angular_length / 2));
+    circle_angular_->clear();
+    circle_angular_->setLineWidth(width_ * 0.05);
+    for (int i = 4; i <= 32; i++) {
```

Comment:
Can you maybe explain where these numbers come from, either in a comment or by making them constants somewhere?

Reply:
This code is migrated from rviz1, let me take a deep look

Reply:
This comment is still open.